src/meta/meta_split_service.cpp

```diff
@@ -293,5 +293,165 @@ void meta_split_service::on_add_child_on_remote_storage_reply(error_code ec,
     parent_context.stage = config_status::not_pending;
 }
 
+void meta_split_service::control_partition_split(control_split_rpc rpc)
+{
+    const auto &req = rpc.request();
+    const auto &control_type = req.control_type;
+    auto &response = rpc.response();
+
+    zauto_write_lock l(app_lock());
+    std::shared_ptr<app_state> app = _state->get_app(req.app_name);
+    if (app == nullptr || app->status != app_status::AS_AVAILABLE) {
+        response.err = app == nullptr ? ERR_APP_NOT_EXIST : ERR_APP_DROPPED;
+        response.__set_hint_msg(fmt::format(
+            "app {}", response.err == ERR_APP_NOT_EXIST ? "not existed" : "dropped", req.app_name));
+        derror_f("{} split failed, {}", control_type_str(control_type), response.hint_msg);
+        return;
+    }
+
+    if (app->helpers->split_states.splitting_count <= 0) {
+        response.err = ERR_INVALID_STATE;
+        response.__set_hint_msg(fmt::format("app({}) is not splitting", req.app_name));
+        derror_f("{} split failed, {}", control_type_str(control_type), response.hint_msg);
+        return;
+    }
+
+    if (req.parent_pidx >= 0 && (control_type == split_control_type::PSC_PAUSE ||
+                                 control_type == split_control_type::PSC_RESTART)) {
+        do_control_single(std::move(app), std::move(rpc));
+    } else {
+        do_control_all(std::move(app), std::move(rpc));
+    }
+}
+
+void meta_split_service::do_control_single(std::shared_ptr<app_state> app, control_split_rpc rpc)
+{
+    const auto &req = rpc.request();
+    const std::string &app_name = req.app_name;
+    const int32_t &parent_pidx = req.parent_pidx;
+    const auto &control_type = req.control_type;
+    auto &response = rpc.response();
+
+    if (parent_pidx >= app->partition_count / 2) {
+        response.err = ERR_INVALID_PARAMETERS;
+        response.__set_hint_msg(fmt::format("invalid parent partition index({})", parent_pidx));
+        derror_f("{} split for app({}) failed, {}",
+                 control_type_str(control_type),
+                 app_name,
+                 response.hint_msg);
+        return;
+    }
+
+    auto iter = app->helpers->split_states.status.find(parent_pidx);
+    if (iter == app->helpers->split_states.status.end()) {
+        response.err = control_type == split_control_type::PSC_PAUSE ? ERR_CHILD_REGISTERED
```

Comment:
Why return different err code?  shoudn't it be all `ERR_INVALID_STATE`?

Reply:
If user tries to pause split has already finished, ERR_CHILD_REGISTERED may be better to descibe split is finished.

Reply:
it's also ok(actually the hint message is enough)